File: 611-valid-triangle-number/ken/611.py
class Solution(object):
    def triangleNumber(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """

        nums.sort()

        cnt = 0
        for i in xrange(len(nums) - 2):
            if nums[i] == 0:
                continue

            k = i + 2
            for j in xrange(i + 1, len(nums) - 1):
                while k < len(nums) and nums[i] + nums[j] > nums[k]:
                    k += 1
                k -= 1
                cnt += k - j

        return cnt

File: 611-valid-triangle-number/leochang/611.cpp
class Solution {
public:
    int triangleNumber(vector<int>& nums) {
        if (nums.size() < 3) return 0;
        vector<int> snums(nums);
        int ans = 0;
        sort(nums.begin(), nums.end());
        
        for(int i = 0; i < nums.size()-2; i++) {
            for(int j = i+1, k = i+2; k < nums.size(); k++) {
                while(j < k && nums[i]+nums[j] <= nums[k]) j++;
                ans += k -j;
            }
        }
        
        return ans;
    }
};
